chunky/src/java/se/llbit/chunky/resources/FloatingPointCompressor.java

```diff
@@ -0,0 +1,215 @@
+        smallBuffer[bytesWritten] = (byte) (xoredValue >>> (byteNo * 8));
+        ++bytesWritten;
+      }
+
+      return header;
+    }
+
+    public void encodePair(double first, double second, BufferedOutputStream output) throws IOException {
+      bytesWritten = 1;
+      byte headerFirst = encodeSingle(first);
+      byte headerSecond = encodeSingle(second);
+      byte header = (byte) ((headerFirst << 4) | headerSecond);
+      smallBuffer[0] = header;
+
+      output.write(smallBuffer, 0, bytesWritten);
+    }
+
+    public void encodeSingleWithOddTerminator(double val, BufferedOutputStream output) throws IOException {
+      bytesWritten = 1;
+      byte headerFirst = encodeSingle(val);
+      byte headerSecond = 0b1100; // Pretend we have 7 zero bytes but the 8th byte is 0 as well
+      smallBuffer[bytesWritten] = 0;
+      ++bytesWritten;
+      byte header = (byte) ((headerFirst << 4) | headerSecond);
+      smallBuffer[0] = header;
+
+      output.write(smallBuffer, 0, bytesWritten);
+    }
+
+    public double decodeSingle(byte header, BufferedInputStream input) throws IOException {
+      long prediction = predictFcm();
+      long dfcmPrediction = predictDfcm();
+
+      byte choiceBit = (byte) (header & 1);
+      byte encodedZeroBytes = (byte) (header >>> 1);
+
+      if(choiceBit == 1)
+        prediction = dfcmPrediction;
+
+      int zeroBytes = encodedZeroBytes;
+      if(zeroBytes > 3)
+        ++zeroBytes;
+
+      long difference = 0;
+      int byteToRead = 8 - zeroBytes;
+      for(int i = 0; i < byteToRead; ++i) {
+        difference = (difference << 8) | input.read();
+      }
+
+      long bits = prediction ^ difference;
+      current = bits;
+
+      return Double.longBitsToDouble(bits);
+    }
+  }
+
+  public static void compress(double[] input, OutputStream output) throws IOException {
+    try(BufferedOutputStream out = new BufferedOutputStream(output)) {
+      if(input.length % 3 != 0)
+        throw new RuntimeException("Dump doesn't have a multiple of 3 values");
+      int pixels = input.length/3;
+      int size = pixels-1;
+      EncoderDecoder rEncoder = new EncoderDecoder();
+      EncoderDecoder gEncoder = new EncoderDecoder();
+      EncoderDecoder bEncoder = new EncoderDecoder();
+      for(int i = 0; i < size; i += 2) {
+        int idx = 3*i;
+        rEncoder.encodePair(input[idx], input[idx+3], out);
+        gEncoder.encodePair(input[idx+1], input[idx+4], out);
+        bEncoder.encodePair(input[idx+2], input[idx+5], out);
+      }
+
+      // Add the last one and a special terminator if there is an odd number
+      if(pixels % 2 == 1) {
+        int idx = 3*size;
+        rEncoder.encodeSingleWithOddTerminator(input[idx], out);
+        gEncoder.encodeSingleWithOddTerminator(input[idx+1], out);
+        bEncoder.encodeSingleWithOddTerminator(input[idx+2], out);
+      }
+    }
+  }
+
+  public static void decompress(double[] output, InputStream input) throws IOException {
```

Comment:
(nit) For the `compress` method, it's `input, output`. Should the arguments be switched here?

Reply:
Yeah, I let like that to get the types in the same order but I don't really care